<a href="https://colab.research.google.com/github/brendanpshea/programming_problem_solving/blob/main/Programming_Part2_Review.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Exceptions
In Python, exceptions are events detected during execution that interrupt the normal flow of a program. The language provides a robust mechanism to handle these exceptions gracefully, using try, except, and finally blocks. This mechanism allows a program to respond to different types of errors without crashing abruptly. Let's delve into how this mechanism can be applied in the context of a simple sushi restaurant script.

Consider a script for a sushi restaurant's order system. Various exceptions might occur, such as an order for an item not on the menu, entering a non-integer value for the quantity, or trying to access a file that logs orders but doesn't exist. Handling these exceptions properly ensures the program runs smoothly and provides meaningful feedback to the user or the system.

### Example 1: Handling a Non-existent Menu Item

In [1]:
menu = ["salmon nigiri", "tuna roll", "avocado maki"]

try:
    order = input("Enter your order: ")
    if order not in menu:
        raise ValueError(f"Item {order} not on the menu")
except ValueError as e:
    print(e)
finally:
    print("Thank you for visiting our sushi restaurant.")


Enter your order: tuna
Item tuna not on the menu
Thank you for visiting our sushi restaurant.


In this example, if a customer orders something not on the menu, a `ValueError` is raised, and the except block handles it by printing the error message. The `finally` block executes regardless of the exception, thanking the customer.

### Example 2: Handling Invalid Quantity Input

In [2]:
try:
    quantity = int(input("Enter quantity: "))
except ValueError:
    print("Please enter a valid integer for quantity.")
finally:
    print("Processing completed.")

Enter quantity: 3
Processing completed.


Here, if the user inputs a non-integer value for quantity, a `ValueError` is raised, and the corresponding except block informs the user to provide a valid integer.

## Example 3: File Handling

In [3]:
try:
    with open("order_log.txt", "r") as file:
        orders = file.read()
except FileNotFoundError:
    print("Order log file not found.")
finally:
    print("End of order processing.")

Order log file not found.
End of order processing.


### Table of 10 Common Python Exceptions

| Exception Name | Description |
| --- | --- |
| ValueError | Raised when a built-in operation or function receives an argument with the right type but an inappropriate value. |
| TypeError | Occurs when an operation or function is applied to an object of inappropriate type. |
| IndexError | Raised when attempting to access an index out of the range of a sequence (e.g., a list or tuple). |
| KeyError | Occurs when a dictionary key is not found. |
| FileNotFoundError | Raised when an attempt to open a file (or a specified path) fails. |
| ZeroDivisionError | Occurs when the second argument of a division or modulo operation is zero. |
| NameError | Raised when a local or global name is not found. |
| OverflowError | Occurs when an arithmetic operation exceeds the limits of the data type. |
| SyntaxError | Raised when the parser encounters a syntax error. |
| AttributeError | Occurs when an attribute reference or assignment fails. |

Understanding and handling these exceptions are fundamental in developing resilient Python applications that can anticipate and manage errors gracefully, enhancing both the user experience and system stability.

### General File Handling in Python

File handling is a fundamental aspect of many programming tasks, enabling applications to persist data between sessions, interact with data files, and process user input or outputs. Python simplifies file operations through its built-in functions, allowing for efficient and straightforward file reading, writing, and manipulation. This section introduces the core concepts of file handling in Python, serving as a precursor to more structured data handling methods like CSV and JSON.

#### Opening Files

The `open()` function is the gateway to file handling in Python. It returns a file object and is commonly used with two arguments: `filename` and `mode`.

`file = open("example.txt", "r")`

#### File Modes

-   "r" - Read: Default mode. Opens a file for reading.
-   "w" - Write: Opens a file for writing, creates the file if it does not exist, and truncates the file to zero length if it does.
-   "a" - Append: Opens a file for appending at the end without truncating it. Creates a new file if it does not exist.
-   "r+" - Read/Write: Opens the file for both reading and writing.

#### The `with` Statement

Using `with` is highly recommended when dealing with file operations. It ensures that the file is properly closed after its suite finishes, even if an exception is raised.

```python
with open("example.txt", "r") as file:
    data = file.read()
```

#### Reading Files

-   `read(size)` - Reads at most `size` bytes from the file. If the `size` argument is omitted or negative, the entire content of the file will be read and returned.
-   `readline()` - Reads a single line from the file.
-   `readlines()` - Reads all the lines in a file and returns them as a list.

```python
with open("example.txt", "r") as file:
    for line in file:
        print(line, end='')
```

#### Writing to Files

-   `write(string)` - Writes the string to the file, returning the number of characters written.
-   `writelines(list)` - Writes a list of strings to the file.

```python
with open("example.txt", "w") as file:
    file.write("Hello, world!\n")
```

#### Appending to Files

Appending to a file adds content to the end, preserving the existing data.

```python
with open("example.txt", "a") as file:
    file.write("Appending to the file.\n")
```

## CSV Files

A CSV (Comma-Separated Values) file is a type of plain text file that uses specific structuring to arrange tabular data. Because it's a plain text file, it can be easily read by humans and machines. CSV files are commonly used for exchanging data between different applications. In a CSV file, each line corresponds to a row in the table, and each field (or cell in the table) is separated by a comma or another delimiter like a semicolon. The simplicity of this format makes it a universal standard for data interchange.

Python provides the `csv` module to facilitate reading from and writing to CSV files. Below are basic examples of how to read from and write to CSV files.

### Writing to a CSV File

To create a simple menu for a sushi restaurant, we'll first write this menu to a CSV file.

In [4]:
import csv

menu = [
    ["Sushi Name", "Ingredients", "Price"],
    ["The Godfather Roll", ["Salmon", "avocado", "cream cheese"], 12.5],
    ["Casablanca Sashimi", ["Tuna"], 10.0],
    ["Gone with the Wind Gunkan", ["Scallop", "mayo"], 8.75],
    ["Citizen Kane Nigiri", ["Eel"], 7.5],
    ["The Shawshank Temaki", ["Crab", "cucumber", "avocado"], 9.25],
    ["Pulp Fiction Uramaki", ["Shrimp tempura", "avocado", "cucumber"], 11.0],
    ["Forrest Gump Futomaki", ["Pickled radish"], 8.0],
    ["Inception Roll", ["Lobster", "caviar", "asparagus"], 18.5],
    ["Interstellar Maki", ["Spicy tuna", "jalapeno"], 13.0],
    ["La La Land Nigiri", ["Salmon belly"], 14.75],
    ["Mad Max Roll", ["Beef", "scallion", "asparagus"], 15.5],
    ["The Matrix Uramaki", ["Avocado", "cucumber", "tofu"], 10.25],
    ["No Country for Old Men Temaki", ["Tuna", "spicy mayo"], 12.0],
    ["Once Upon a Time Temaki", ["Shrimp", "mango", "cream cheese"], 13.25],
    ["Parasite Gunkan", ["Sea urchin"], 19.0],
    ["The Silence of the Lambs Sashimi", ["Octopus"], 11.5],
    ["Skyfall Nigiri", ["Smoked salmon"], 9.75],
    ["Titanic Roll", ["Cucumber", "avocado", "salmon"], 16.0],
    ["Whiplash Maki", ["Eel", "avocado"], 10.5],
]

try:
    with open('sushi_menu.csv', 'w', newline='') as file:
        writer = csv.writer(file)
        writer.writerows(menu)
    print("Menu successfully written to sushi_menu.csv")
except IOError:
    print("Error writing to file.")


Menu successfully written to sushi_menu.csv


In [5]:
# Open file using system viewer
!cat sushi_menu.csv

Sushi Name,Ingredients,Price
The Godfather Roll,"['Salmon', 'avocado', 'cream cheese']",12.5
Casablanca Sashimi,['Tuna'],10.0
Gone with the Wind Gunkan,"['Scallop', 'mayo']",8.75
Citizen Kane Nigiri,['Eel'],7.5
The Shawshank Temaki,"['Crab', 'cucumber', 'avocado']",9.25
Pulp Fiction Uramaki,"['Shrimp tempura', 'avocado', 'cucumber']",11.0
Forrest Gump Futomaki,['Pickled radish'],8.0
Inception Roll,"['Lobster', 'caviar', 'asparagus']",18.5
Interstellar Maki,"['Spicy tuna', 'jalapeno']",13.0
La La Land Nigiri,['Salmon belly'],14.75
Mad Max Roll,"['Beef', 'scallion', 'asparagus']",15.5
The Matrix Uramaki,"['Avocado', 'cucumber', 'tofu']",10.25
No Country for Old Men Temaki,"['Tuna', 'spicy mayo']",12.0
Once Upon a Time Temaki,"['Shrimp', 'mango', 'cream cheese']",13.25
Parasite Gunkan,['Sea urchin'],19.0
The Silence of the Lambs Sashimi,['Octopus'],11.5
Skyfall Nigiri,['Smoked salmon'],9.75
Titanic Roll,"['Cucumber', 'avocado', 'salmon']",16.0
Whiplash Maki,"['Eel', 'avocado']",10.5


This example tries to write the extended menu to a CSV file. If an error occurs during file operations (e.g., permission issues), it is caught by the `except` block, which prints an error message.

### Reading a CSV file

In [6]:
import csv

try:
    with open('sushi_menu.csv', 'r') as file:
        reader = csv.reader(file)
        for row in reader:
            print(row)
except FileNotFoundError:
    print("The file sushi_menu.csv was not found.")
except IOError:
    print("Error reading file.")

['Sushi Name', 'Ingredients', 'Price']
['The Godfather Roll', "['Salmon', 'avocado', 'cream cheese']", '12.5']
['Casablanca Sashimi', "['Tuna']", '10.0']
['Gone with the Wind Gunkan', "['Scallop', 'mayo']", '8.75']
['Citizen Kane Nigiri', "['Eel']", '7.5']
['The Shawshank Temaki', "['Crab', 'cucumber', 'avocado']", '9.25']
['Pulp Fiction Uramaki', "['Shrimp tempura', 'avocado', 'cucumber']", '11.0']
['Forrest Gump Futomaki', "['Pickled radish']", '8.0']
['Inception Roll', "['Lobster', 'caviar', 'asparagus']", '18.5']
['Interstellar Maki', "['Spicy tuna', 'jalapeno']", '13.0']
['La La Land Nigiri', "['Salmon belly']", '14.75']
['Mad Max Roll', "['Beef', 'scallion', 'asparagus']", '15.5']
['The Matrix Uramaki', "['Avocado', 'cucumber', 'tofu']", '10.25']
['No Country for Old Men Temaki', "['Tuna', 'spicy mayo']", '12.0']
['Once Upon a Time Temaki', "['Shrimp', 'mango', 'cream cheese']", '13.25']
['Parasite Gunkan', "['Sea urchin']", '19.0']
['The Silence of the Lambs Sashimi', "['Octopus

In this reading example, we specifically catch `FileNotFoundError` to handle the case where the file does not exist. A general `IOError` is also caught to cover other I/O related errors. This ensures that the program provides a clear message about what went wrong instead of terminating abruptly.

## Table: Reading and Writing CSV Files

| Code Snippet | Explanation |
| --- | --- |
| `with open("file_path", "r") as file:` | Open a file for reading. The `with` statement ensures proper resource management, automatically closing the file after the block's execution. |
| `with open("file_path", "w") as file:` | Open a file for writing. If the file exists, it will be overwritten. If the file doesn't exist, it will be created. |
| `with open("file_path", "a") as file:` | Open a file for appending. If the file exists, data will be appended to the end. If the file doesn't exist, it will be created. |
| `file.read()` | Read the entire contents of a file into a string. |
| `file.readline()` | Read the next line from the file, including the newline character. |
| `file.readlines()` | Read all lines in a file into a list where each element is a line. |
| `file.write("text")` | Write the string "text" to the file. |
| `import csv` | Import the CSV module, which provides functionality to read and write CSV files. |
| `csv.reader(file)` | Create a CSV reader object which will iterate over lines in the specified file. |
| `csv.writer(file)` | Create a CSV writer object which provides methods to write to the CSV file. |
| `writer.writerow(["data1", "data2"])` | Write a single row to the CSV file. The argument is a list where each element corresponds to a field. |
| `writer.writerows([["row1data1", "row1data2"], ["row2data1", "row2data2"]])` | Write multiple rows to the CSV file. The argument is a list of lists, with each inner list representing a row. |
| `reader = csv.DictReader(file)` | Create a CSV reader object that maps the information read into a dictionary where keys are given by the optional `fieldnames` parameter or the first row in the file. |
| `writer = csv.DictWriter(file, fieldnames=["col1", "col2"])` | Create a CSV writer object to write dictionaries to the file, using the `fieldnames` parameter to specify the column names. |

## JSON Files

JSON (JavaScript Object Notation) is a lightweight data-interchange format that is easy for humans to read and write and easy for machines to parse and generate. JSON is based on the syntax of JavaScript but is a text format that is completely language independent. JSON files are ideal for data interchange between a server and a web application or between different parts of an application. A JSON file typically has a `.json` extension and represents data structured in a hierarchy of objects and arrays. Objects are denoted by curly braces `{}` containing key-value pairs, where the key is a string, and the value can be a string, number, boolean, array, or another object. Arrays are ordered lists of values and are denoted by square brackets `[]`.

Python comes with a built-in module named `json` for encoding and decoding JSON data. You can easily convert between Python dictionaries and JSON formatted strings, and vice versa.

#### Writing to a JSON File
To write a Python dictionary to a file:

In [7]:
import json

# Data to be written
sushi_menu = {
    "rolls": [
        {"name": "The Matrix Reloaded Roll", "ingredients": ["Avocado", "cucumber", "green onion"], "price": 9.5},
        {"name": "The Incredibles Roll", "ingredients": ["Tempura shrimp", "spicy tuna", "avocado"], "price": 12.5},
        # Add more rolls here
    ]
}

# Writing JSON data
with open('sushi_menu.json', 'w') as file:
    json.dump(sushi_menu, file, indent=4)


In [8]:
# View the file with the system viewer
!cat sushi_menu.json

{
    "rolls": [
        {
            "name": "The Matrix Reloaded Roll",
            "ingredients": [
                "Avocado",
                "cucumber",
                "green onion"
            ],
            "price": 9.5
        },
        {
            "name": "The Incredibles Roll",
            "ingredients": [
                "Tempura shrimp",
                "spicy tuna",
                "avocado"
            ],
            "price": 12.5
        }
    ]
}

To read a JSON file back into JSON, we can do the following:

In [9]:
import json

# Reading JSON data
with open('sushi_menu.json', 'r') as file:
    menu_data = json.load(file)

print(menu_data)

{'rolls': [{'name': 'The Matrix Reloaded Roll', 'ingredients': ['Avocado', 'cucumber', 'green onion'], 'price': 9.5}, {'name': 'The Incredibles Roll', 'ingredients': ['Tempura shrimp', 'spicy tuna', 'avocado'], 'price': 12.5}]}


## Table: CSV and JSON

| Feature | JSON | CSV |
| --- | --- | --- |
| Format | Text-based, hierarchical | Text-based, tabular |
| Data Structure | Supports nested and complex data. <br> Ideal for hierarchical or structured data with varying fields. | Flat structure.  <br> Best for simple tabular data without hierarchical relationships. |
| Readability | Readable by humans and easily parsed by machines.  <br> Provides clear visibility of nested data. | Easily readable by humans, especially in spreadsheet programs. <br> Less clear for complex hierarchical data. |
| Flexibility | Highly flexible in representing different data types and structures. | Limited flexibility. Best suited for straightforward, table-like data. |
| Standardization | Widely used in web applications for data interchange.  <br>JSON format is universally recognized across programming languages. | A common format for data export and import in spreadsheets and databases.  <br>Less standardized in terms of data interpretation. |
| File Size | Generally larger due to repeated keys and more verbose formatting.  <br>Can be minimized with compression. | Typically smaller and more compact, making it efficient for large datasets. |
| Ease of Use | Easy to use with modern web APIs and JavaScript-based applications. <br> Requires parsing to and from objects in most programming languages. | Simple to use with minimal overhead for reading and writing. <br> Directly compatible with spreadsheet applications. |
| Data Types | Supports strings, numbers, arrays, and Boolean values directly. | Primarily treats all data as strings. <br> Requires explicit conversion for numerical operations. |
| Use Cases | Ideal for applications that require complex data with nested structures, <br> such as configurations, web API responses, and settings. | Suitable for data analysis, spreadsheet manipulation, <br> and scenarios where data is uniformly structured without a need for nesting. |

## Sample Program: Random Rolls
Here is a sample program that demonstrates some of what we've been coverning so far. It creates a menu of random sushi rolls.

In [2]:
import random

def random_rolls(n):
    """ A function to create n random sushi rolls"""
    ingredients = [
        ("salmon", 60), ("tuna", 50), ("red snapper", 60), ("avocado", 160), ("cucumber", 8),
        ("eel", 85), ("shrimp", 20), ("crab meat", 40), ("tempura", 95), ("mayo", 100),
        ("spicy mayo", 110), ("tofu", 70), ("cream cheese", 100), ("sweet potato", 90),
        ("asparagus", 20), ("scallions", 5), ("mango", 70), ("red bell pepper", 25),
        ("carrot", 40), ("pickled radish", 15)
    ]
    rolls = []

    for _ in range(n):
        roll_type = random.choice(["maki", "nigiri"])
        num_ingredients = random.randint(2, 4)
        selected_ingredients = random.sample(ingredients, num_ingredients)
        ingredients_names = [ing[0] for ing in selected_ingredients]
        total_calories = sum(ing[1] for ing in selected_ingredients)

        # For nigiri, add rice calories
        if roll_type == "nigiri":
            total_calories += 40  # Assuming rice adds 40 calories
        # For maki, add rice and seaweed calories
        else:
            total_calories += 50  # Assuming rice and seaweed together add 50 calories

        sushi_name = " and ".join(ingredients_names) + f" {roll_type}"
        rolls.append((sushi_name, ingredients_names, total_calories, roll_type))

    return rolls

# Example of generating 5 random sushi rolls
print(random_rolls(5))

[('eel and pickled radish maki', ['eel', 'pickled radish'], 150, 'maki'), ('asparagus and cream cheese and mayo nigiri', ['asparagus', 'cream cheese', 'mayo'], 260, 'nigiri'), ('cream cheese and red snapper nigiri', ['cream cheese', 'red snapper'], 200, 'nigiri'), ('asparagus and mango and scallions and cream cheese maki', ['asparagus', 'mango', 'scallions', 'cream cheese'], 245, 'maki'), ('eel and tuna nigiri', ['eel', 'tuna'], 175, 'nigiri')]


Here's what happens here:

1.  `def random_rolls(n):` defines a function named `random_rolls` that accepts a single argument, `n`, indicating the number of sushi rolls to generate.

2.  `selected_ingredients = random.sample(ingredients, num_ingredients)` selects a random sample of ingredients from the `ingredients` list.

3.  `ingredients_names = [ing[0] for ing in selected_ingredients]` uses list comprehension to extract the first element (name) from each tuple in the `selected_ingredients` list.

4.  `total_calories = sum(ing[1] for ing in selected_ingredients)` calculates the total caloric value of the selected ingredients using aggregation with the `sum()` function, combined with a "generator expression".

5.  `rolls.append((sushi_name, ingredients_names, total_calories, roll_type))` adds a tuple containing the sushi roll's name, its ingredients, the total caloric value, and the roll type to the `rolls` list.

Now, let's see how we can generate 100 rolls and write these to a CSV file.

In [4]:
import csv

def write_sushi_rolls_to_csv(sushi_rolls, filename="sushi_rolls.csv"):
    headers = ["Sushi Name", "Ingredients", "Total Calories", "Type"]

    try:
        # Open file in write mode.
        with open(filename, 'w') as file:
            writer = csv.writer(file)
            writer.writerow(headers)  # Write the header row
            for roll in sushi_rolls:
                # Prepare the ingredients list as a string for CSV
                ingredients_str = ', '.join(roll[1])
                writer.writerow([roll[0], ingredients_str, roll[2], roll[3]])

        print(f"Sushi rolls data successfully written to {filename}")
    except IOError as e:
        print(f"Failed to write to file {filename}. Error: {e}")
    except Exception as e:
        print(f"An unexpected error occurred: {e}")

sushi_rolls = random_rolls(100)
write_sushi_rolls_to_csv(sushi_rolls)

Sushi rolls data successfully written to sushi_rolls.csv


A few important lines here:

1.  `def write_sushi_rolls_to_csv(sushi_rolls, filename="sushi_rolls.csv")` establishes a method to save sushi roll information into a CSV file, accepting the data and an optional file name. We specify a default value for the filename.

2.  `with open(filename, 'w') as file:` opens a file for writing. The indented lines that follow can interact with this file. Once we "un-indent", Python will automatically close the file.

3.  `writer = csv.writer(file)` initializes a CSV writer object, preparing it to convert and store the provided data into the specified file.

4.  `writer.writerow(headers)` inputs column names at the top of the CSV file, laying out the structure for the data entries that follow.

5.  `writer.writerow([roll[0], ingredients_str, roll[2], roll[3]])` iterates over each sushi roll's data, formatting and recording it into the CSV file according to the predefined headers.

6. We use `try` and `except` blocks to handle possible problems with writing to the file.

## Exercise: Create a Random Pizza Generator in Python

The goal of this exercise is to adapt the given sushi roll generator functions to create a random pizza generator. This generator will produce a list of pizzas with randomly selected toppings, a type of crust, and calculate the total calories. This can be closely based on the above sushi generator app, and you should be able to reuse much of the same code.

#### Step 1: Define Pizza Ingredients

Create a list of tuples for pizza ingredients, similar to the sushi ingredients. Each tuple should contain the ingredient name and its calorie count. Consider ingredients like various cheeses, meats, vegetables, and sauces.

#### Step 2: Implement the Random Pizza Generator Function
Adapt the `random_rolls` function to generate random pizzas. Each pizza should have a random number of toppings (e.g., 2 to 5), a type of crust (e.g., thin, thick, stuffed), and calculate the total calories. Generate a name for each pizza based on its toppings.

#### Step 3: Modify the Output Structure
Modify the structure of the output tuple to fit the pizza context. It could include the pizza name, list of toppings, total calories, and crust type.

#### Step 4: Write the Pizza Data to a CSV File

Adapt the `write_sushi_rolls_to_csv` function to write your pizza data to a CSV file. Ensure your file includes headers appropriate for pizza (e.g., "Pizza Name", "Toppings", "Total Calories", "Crust Type").


In [ ]:
def pizza_generator(n):
  # Your code here
  pass

In [ ]:
def write_pizza_to_csv(sushi_rolls, filename="pizzas.csv"):
  # Your code here
  pass

## What is pandas?

pandas is an open-source data analysis and manipulation tool, built on top of the Python programming language. It offers data structures and operations for manipulating numerical tables and time series, making it one of the most preferred and widely used tools in data analytics. pandas is particularly suited for working with tabular data (similar to Excel spreadsheets), including data that might have heterogeneous types, missing values, or other complexities.
Some basic concepts of Pandas include:

-  A **Data Series** is a one-dimensional array-like object containing a sequence of values (similar to a list in Python) and an associated array of data labels, called its index. Think of it as a single column of a spreadsheet, with row labels.

-   A **Data Frame** is a two-dimensional, size-mutable, and potentially heterogeneous tabular data structure with labeled axes (rows and columns). It can be thought of as a spreadsheet or SQL table, or a dict of Series objects. Data Frames are the most commonly used pandas object, facilitating the handling of data in a structured form.

pandas provides built in methods that are focused on the following tasks:

-   **Data Cleaning.** pandas provides various functions for easily dropping missing values, filling missing values with specific data, and finding duplicates, which are essential steps in preparing data for analysis.

-   **Data Analytics.** With pandas, you can perform both simple and complex data analysis tasks. These include aggregations, summarizations, transformations, and more sophisticated operations like group by, merge, join, and time series analysis.

-   **Data Visualization.** pandas also integrates with Matplotlib for basic plotting functionality, allowing for quick and easy visualization of the data contained within a DataFrame or Series, including line plots, bar plots, histograms, and scatter plots.

To utilize pandas for working with our sushi rolls data, we would start by loading the `sushi_rolls.csv` file into a pandas DataFrame. This process converts the structured CSV file into a DataFrame, enabling us to perform more complex manipulations and analyses with greater ease.

In [6]:
import pandas as pd
sushi_df = pd.read_csv("sushi_rolls.csv")
sushi_df.head()

Sushi Name  \
0                      sweet potato and shrimp maki   
1                   avocado and pickled radish maki   
2                      spicy mayo and carrot nigiri   
3  crab meat and scallions and shrimp and tuna maki   
4                    cucumber and cream cheese maki   

                          Ingredients  Total Calories    Type  
0                sweet potato, shrimp             160    maki  
1             avocado, pickled radish             225    maki  
2                  spicy mayo, carrot             190  nigiri  
3  crab meat, scallions, shrimp, tuna             165    maki  
4              cucumber, cream cheese             158    maki

### Filtering, Selecting, and Sorting Data in pandas

Filtering, selecting, and sorting data are foundational operations that enable analysts to focus on specific subsets of data, identify patterns, and organize data in meaningful ways. These operations form the core of data manipulation tasks, allowing for the extraction of relevant information from larger datasets.

-   Filtering involves narrowing down the data to a subset that meets certain criteria. This could be rows that have specific values in one or more columns.
-   Selecting refers to choosing specific columns or rows from the dataset based on their labels, indices, or conditions.
-   Sorting is the process of ordering the data in ascending or descending order based on one or more columns, which helps in understanding the distribution and relationship of data.


Some sample operations:

| pandas Code | Explanation |
| --- | --- |
| `sushi_df['Total Calories'] < 200` |  Returns a Boolean Series (True or False) of whether rows are less than 200 calories.. |
| `sushi_df[sushi_df['Type'] == 'maki']` |  Selects rows where the 'Type' column is equal to 'maki'. |
| `sushi_df[['Sushi Name', 'Total Calories']]` | Returns a DataFrame containing only the 'Sushi Name' and 'Total Calories' columns. |
| `sushi_df.iloc[0:3]` | Selects the first three rows of the DataFrame using integer indices. |
| `sushi_df.loc[sushi_df['Sushi Name'].str.contains('avocado')]` | Filter rows containing 'avocado' in the Sushi Name. |
| `sushi_df.sort_values(by='Total Calories')` | Sort the DataFrame by Total Calories. |
| `sushi_df.sort_values(by=['Type', 'Total Calories'], ascending=[True, False])` | Sorts the DataFrame first by 'Type' in ascending order, then by 'Total Calories' in descending order within each 'Type'. |
| `sushi_df[sushi_df['Ingredients'].str.split(', ').apply(len) > 2]` | Filter rows with more than two ingredients by applying the `len()` function to each row. |
| `sushi_df.drop_duplicates(subset='Sushi Name')` | Drop duplicate rows based on the 'Name' column. |
| `sushi_df.groupby('Type')[["Total Calories"]].mean()` | Group by 'Type' and calculate mean calories. |

In [19]:
# Try these out here!

### Creating New columns
In pandas, creating new columns in a DataFrame is a common operation that allows for the enrichment of the dataset with additional information or metrics derived from existing data. Creating new columns can serve multiple purposes:

-   Generate new data points from existing columns, such as ratios, sums, or averages.
-   Classify rows into categories based on certain conditions or thresholds.
-   Apply transformations or operations to existing data to create a new representation, such as scaling, normalization, or string manipulation.

A few sample operations:

| Code | Description |
| --- | --- |
| `sushi_df['Ingredient Count'] = sushi_df['Ingredients'].str.split(', ').apply(len)` | Creates a new column called 'Ingredient Count' that stores the number of ingredients in each sushi roll. |
| `sushi_df['Contains Avocado'] = sushi_df['Ingredients'].str.contains('avocado')` |Adds a boolean column 'Contains Avocado' to indicate whether the sushi roll includes avocado. |
| `sushi_df['Calories per Ingredient'] = sushi_df['Total Calories'] / sushi_df['Ingredient Count']` | Generates a new column 'Calories per Ingredient' by dividing the total calories by the ingredient count. |
| `sushi_df['Is Low Calorie'] = sushi_df['Total Calories'] < 150` | Introduces a boolean column 'Is Low Calorie' to flag sushi rolls with less than 150 total calories. |
| `sushi_df['Type Code'] = sushi_df['Type'].map({'maki': 0, 'nigiri': 1})` |  Adds a numeric column 'Type Code' by mapping 'maki' to 0 and 'nigiri' to 1. |
| `sushi_df['First Ingredient'] = sushi_df['Ingredients'].str.split(', ').str[0]` | Creates a new column 'First Ingredient' to store the first ingredient listed for each sushi roll. |